frost/signer.go

```diff
@@ -364,3 +364,87 @@ func (s *Signer) encodeGroupCommitment(commitments []*NonceCommitment) []byte {
 	// return encoded_group_commitment
 	return b
 }
+
+// deriveInterpolatingValue implements def derive_interpolating_value(L, x_i)
+// function from [FROST], as defined in section 4.2 Polynomials.
+// L is the list of the indices of the members of the particular group.
+// xi is the index of the participant i.
+func (s *Signer) deriveInterpolatingValue(xi uint64, L []uint64) (*big.Int, error) {
+	// From [FROST]:
+	//
+	// 4.2.  Polynomials
+	//
+	//   This section defines polynomials over Scalars that are used in the
+	//   main protocol.  A polynomial of maximum degree t is represented as a
+	//   list of t+1 coefficients, where the constant term of the polynomial
+	//   is in the first position and the highest-degree coefficient is in the
+	//   last position.  For example, the polynomial x^2 + 2x + 3 has degree 2
+	//   and is represented as a list of 3 coefficients [3, 2, 1].  A point on
+	//   the polynomial f is a tuple (x, y), where y = f(x).
+	//
+	//   The function derive_interpolating_value derives a value used for
+	//   polynomial interpolation.  It is provided a list of x-coordinates as
+	//   input, each of which cannot equal 0.
+	//
+	//   Inputs:
+	//     - L, the list of x-coordinates, each a NonZeroScalar.
+	//     - x_i, an x-coordinate contained in L, a NonZeroScalar.
+	//
+	//   Outputs:
+	//     - value, a Scalar.
+	//
+	//   Errors:
+	//     - "invalid parameters", if 1) x_i is not in L, or if 2) any
+	//       x-coordinate is represented more than once in L.
+	//
+	//   def derive_interpolating_value(L, x_i):
+
+	order := s.ciphersuite.Curve().Order()
+	found := false
+	// numerator = Scalar(1)
+	num := big.NewInt(1)
+	// denominator = Scalar(1)
+	den := big.NewInt(1)
+	// for x_j in L:
+	for _, xj := range L {
+		if xj == xi {
+			// for x_j in L:
+			//     if count(x_j, L) > 1:
+			//         raise "invalid parameters"
+			if found {
```

Comment:
This doesn't actually implement the specified functionality. We need to expect that the indices `L` are sorted in ascending order, with no repeated instances, and keep track of the previous `x_j'`. Comparing `x_j` to `x_j'` and requiring that `x_j > x_j'` gives the specified functionality. This was a genuine oversight in the prototype.

Reply:
That's true though looking at `Round2` implementation in #8 I do wonder if we need to worry about it in `deriveInterpolatingValue` at all. `validateGroupCommitments` makes sure that all commitments are sorted in ascending order and there is no duplicates.

We could improve unit tests for `validateGroupCommitments` to cover duplicates and make it clear in `deriveInterpolatingValue` that the validation must happen before this function is called.

This would allow us to eliminate this error handling:
https://github.com/threshold-network/roast-go/blob/579570232aa2bd85881be676eec5dd47dceade5f/frost/signer.go#L120-L125

and `validateGroupCommitments` would be the only function returning errors in `Round2`. This wouldn't be 1:1 with the paper but this solution feels quite elegant to me.

What do you think about it?

Reply:
Regardless of the decision here, I added a test case to `validateGroupCommitments` unit tests covering the duplicate element case: 1398d05c9b025927f46203b0f6f2608526946e97.

Reply:
I think if we note in a comment that `validateGroupCommitments` is a necessary part for checking the correctness, it should be fine.

Reply:
I will handle it in #8.